File: MATTHEW/leetcode/AIcoding/4.GPT-4o/cpp_solutions/1123-lowest-common-ancestor-of-deepest-leaves.cpp
class Solution {
public:
    TreeNode* lcaDeepestLeaves(TreeNode* root) {
        int maxDepth = maxDepthOfTree(root);
        return lcaHelper(root, 1, maxDepth);
    }

private:
    int maxDepthOfTree(TreeNode* node) {
        if (!node) return 0;
        return 1 + max(maxDepthOfTree(node->left), maxDepthOfTree(node->right));
    }

    TreeNode* lcaHelper(TreeNode* node, int depth, int maxDepth) {
        if (!node) return nullptr;
        if (depth == maxDepth) return node;

        TreeNode* left = lcaHelper(node->left, depth + 1, maxDepth);
        TreeNode* right = lcaHelper(node->right, depth + 1, maxDepth);

        if (left && right) return node;
        return left ? left : right;
    }
};

File: leetcode/AIcoding/5.Qwen-3/java_solutions/1123-lowest-common-ancestor-of-deepest-leaves.java
public class Solution {
    public TreeNode lcaDeepestLeaves(TreeNode root) {
        return findLCA(root, 0).node;
    }

    private static class Pair {
        TreeNode node;
        int depth;

        Pair(TreeNode node, int depth) {
            this.node = node;
            this.depth = depth;
        }
    }

    private Pair findLCA(TreeNode node, int depth) {
        if (node == null) {
            return new Pair(null, -1);
        }

        Pair left = findLCA(node.left, depth + 1);
        Pair right = findLCA(node.right, depth + 1);

        if (left.depth == right.depth) {
            return new Pair(node, left.depth);
        } else if (left.depth > right.depth) {
            return left;
        } else {
            return right;
        }
    }
}